servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpMessageDiscardWatchdogClientFilter.java

```diff
@@ -51,51 +52,76 @@ final class HttpMessageDiscardWatchdogClientFilter implements StreamingHttpConne
+    static final StreamingHttpConnectionFilterFactory INSTANCE = USE_PHANTOM_REFERENCE ?
+            new PhantomReferenceConnectionFilterFactory() : new AtomicRefFilter();
 
     /**
      * Instance of {@link StreamingHttpClientFilterFactory} with the cleaner implementation.
      */
-    static final StreamingHttpClientFilterFactory CLIENT_CLEANER = new CleanerStreamingHttpClientFilterFactory();
+    static final StreamingHttpClientFilterFactory CLIENT_CLEANER = USE_PHANTOM_REFERENCE ?
+            new NoopCleanerFilterFactory() : new CleanerStreamingHttpClientFilterFactory();
 
     private HttpMessageDiscardWatchdogClientFilter() {
-        // Singleton
+        // no instances
     }
 
-    @Override
-    public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
-        return new StreamingHttpConnectionFilter(connection) {
-            @Override
-            public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
-                return delegate().request(request).map(response -> {
-                    // always write the buffer publisher into the request context. When a downstream subscriber
-                    // arrives, mark the message as subscribed explicitly (having a message present and no
-                    // subscription is an indicator that it must be freed later on).
-                    final AtomicReference<Publisher<?>> reference = request.context()
-                            .computeIfAbsent(MESSAGE_PUBLISHER_KEY, key -> new AtomicReference<>());
-                    assert reference != null;
-                    if (reference.getAndSet(response.messageBody()) != null) {
-                        // If a previous message exists, the Single<StreamingHttpResponse> got resubscribed to
-                        // (i.e. during a retry) and so previous message body needs to be cleaned up by the
-                        // user.
-                        LOGGER.warn("Discovered un-drained HTTP response message body which has " +
-                                "been dropped by user code - this is a strong indication of a bug " +
-                                "in a user-defined filter. Response payload (message) body must " +
-                                "be fully consumed before retrying.");
-                    }
-
-                    return response.transformMessageBody(msgPublisher -> msgPublisher.beforeSubscriber(() -> {
-                        reference.set(null);
-                        return HttpMessageDiscardWatchdogServiceFilter.NoopSubscriber.INSTANCE;
-                    }));
-                });
-            }
-        };
+    // This implementation uses an atomic reference stored on the context to track when requests have been leaked.
+    // It is efficient and doesn't depend on GC, but may be prone to false positives since its possible that the
```

Comment:
As we discussed offline:
`HttpMessageDiscardWatchdogClientFilter` logs only in 2 use-cases:
1. When we detect a new retry but the previous response was not drained. We expect that previous response payload is always drained before the retry is initiated.
2. When we detect `onError` at the outer client filter (close to the user) after we received response meta-data in the inner connection filter (close to channel).

In both cases the previous response is not available anymore and can not be drained later. The false positives are unlikely.

The PhantomReference approach might be helpful to cover more use-cases, like when users use streaming responses, successfully got the response metadata from the client, but then forgot to drain it later in the users scope. However, having 2 different approaches will increase the runtime cost (watchdog already adds visible overhead). On the other hand, removing watchdog and replacing it with PhantomReference at `NettyChannelPublisher` level will not allow us distinguish (1) from (2). Idk... Let's chat more about it on the team sync.